File: src/Shared/Polyrific.Catapult.Shared.ApiClient/AccountService.cs
// Copyright (c) Polyrific, Inc 2018. All rights reserved.

using System.Collections.Generic;
using System.Threading.Tasks;
using Polyrific.Catapult.Shared.Dto.User;
using Polyrific.Catapult.Shared.Service;

namespace Polyrific.Catapult.Shared.ApiClient
{
    public class AccountService : BaseService, IAccountService
    {
        public AccountService(IApiClient api) : base(api)
        {
        }

        public async Task<string> ConfirmEmail(int userId, string token)
        {
            var path = $"account/{userId}/confirm";

            return await Api.Get<string>(path);
        }

        public async Task<UserDto> GetUser(int userId)
        {
            var path = $"account/{userId}";

            return await Api.Get<UserDto>(path);
        }

        public async Task<UserDto> GetCurrentUser()
        {
            var path = $"account/currentuser";

            return await Api.Get<UserDto>(path);
        }

        public async Task<UserDto> GetUserByUserName(string userName)
        {
            var path = $"account/name/{userName}";

            return await Api.Get<UserDto>(path);
        }

        public async Task<UserDto> GetUserByEmail(string email)
        {
            var path = $"account/email/{email}";

            return await Api.Get<UserDto>(path);
        }

        public async Task<List<UserDto>> GetUsers(string status)
        {
            var path = "account";

            return await Api.Get<List<UserDto>>(path);
        }

        public async Task<RegisterUserResultDto> RegisterUser(RegisterUserDto dto)
        {
            var path = "account/register";

            return await Api.Post<RegisterUserDto, RegisterUserResultDto>(path, dto);
        }

        public async Task RemoveUser(int userId)
        {
            var path = $"account/{userId}";

            await Api.Delete(path);
        }

        public async Task SetUserRole(int userId, SetUserRoleDto dto)
        {
            var path = $"account/{userId}/role";

            await Api.Post<SetUserRoleDto, SetUserRoleDto>(path, dto);
        }

        public async Task SuspendUser(int userId)
        {
            var path = $"account/{userId}/suspend";

            await Api.Post<object>(path, null);
        }

        public async Task ReactivateUser(int userId)
        {
            var path = $"account/{userId}/activate";

            await Api.Post<object>(path, null);
        }

        public async Task UpdateUser(int userId, UpdateUserDto dto)
        {
            var path = $"account/{userId}";

            await Api.Put(path, dto);
        }

        public async Task UpdatePassword(int userId, UpdatePasswordDto dto)
        {
            var path = $"account/{userId}/password";

            await Api.Post(path, dto);
        }
    }
}

File: src/Web/opencatapultweb/src/app/core/services/account.service.ts
import { Injectable } from '@angular/core';
import { ApiService } from './api.service';
import { UserDto } from '../models/account/user-dto';
import { SetUserRoleDto } from '../models/account/set-user-role-dto';
import { RegisterUserDto } from '../models/account/register-user-dto';
import { UpdateUserDto } from '../models/account/update-user-dto';
import { UpdatePasswordDto } from '../models/account/update-password-dto';
import { ResetPasswordDto } from '../models/account/reset-password-dto';
import { ExternalAccountTypeDto } from '../models/account/external-account-type-dto';

@Injectable()
export class AccountService {

  constructor(private api: ApiService) { }

  getUserByUserName(userName: string) {
    return this.api.get<UserDto>(`account/name/${userName}`);
  }

  getExternalAccountTypes() {
    return this.api.get<ExternalAccountTypeDto[]>('account/external-type');
  }

  getUsers(status: string, role: string) {
    return this.api.get<UserDto[]>(`account?status=${status}&role=${role}`);
  }

  activate(userId: number) {
    return this.api.postString(`account/${userId}/activate`);
  }

  suspend(userId: number) {
    return this.api.postString(`account/${userId}/suspend`);
  }

  remove(userId: number) {
    return this.api.delete(`account/${userId}`);
  }

  setUserRole(userId: number, dto: SetUserRoleDto) {
    return this.api.post(`account/${userId}/role`, dto);
  }

  register(dto: RegisterUserDto) {
    return this.api.post('account/register', dto);
  }

  updateUser(userId: number, dto: UpdateUserDto) {
    return this.api.put(`account/${userId}`, dto);
  }

  updateAvatar(userId: number, managedFileId: number) {
    return this.api.put(`account/${userId}/avatar?managedFileId=${managedFileId}`, null);
  }

  updatePassword(dto: UpdatePasswordDto) {
    return this.api.putString('account/password', dto);
  }

  requestResetPassword(username: string) {
    return this.api.getString(`account/name/${username}/resetpassword`);
  }

  resetPassword(username: string, dto: ResetPasswordDto) {
    return this.api.postString(`account/name/${username}/resetpassword`, dto);
  }

}
